GoogleCloudPlatform/run-gmp-sidecar
# entrypoint: support hot reloading of configs

This change does the following:
- Watches the user config file for changes and updates the collector config if detected

To test this change do the following:
- Run `./create-sa-and-ar.sh`
- Run `gcloud builds submit . --config=cloudbuild.yaml --region=<SOME REGION>`. Now you have a sample app and a sidecar running.
- Run `gcloud secrets versions add run-gmp-config --data-file=<NEW CONFIG FILE>` to update the secret, where the config is stored.
- Look at metrics and logs, collector should be using updated config.

Notes for reviewer:
- Cloud Run secret mounting doesn't work with fsnotify or polling, so this PR uses `stat` at a 10s interval. I picked 10s arbitrarily - no real science.
- We need better integration tests, but didn't want to block this on that. But we should prioritize that soon.
- We make the collector use the same self metics port on reload so to avoid discontinuity in the self metrics (self metrics port is part of the `instance` label)

Change-Id: I2f15ae2d94e0f94ffe57d9ffb03f883341301ca5
